### PR TITLE
POST request to start session with GovPay

### PR DESCRIPTION
**Pull request by jphillips2121**
Create POST request to start session with GovPay and return JourneyURL to access session.

Resolves: CPS-143

### Type of change

* [X] New feature

### Pull request checklist

* [X] I have added unit tests for new code that I have added
* [X] I have added/updated functional tests where appropriate
* [X] The code follows our [coding standards](https://github.com/companieshouse/styleguides/blob/master/go.md)

__An exhaustive list of peer review checks can be read [here](https://github.com/golang/go/wiki/CodeReviewComments)__

### CODE REVIEW

File: models/gov_pay.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+type NextURLPost struct {
+	PostType string `json:"type"`
+	Params   Params `json:"params"`
+	HREF     string `json:"href"`
+	Method   string `json:"method"`
+}
+
+// Params contains the charge token for the payment
+type Params struct {
+	ChargeTokenId string `json:"chargeTokenId"`
```
Comment (dalebradley) [inline models/gov_pay.go]: Should be ChargeTokenID

File: service/external_paysession.go [hunk 2]
```diff
@@ -4,43 +4,68 @@ import (
...
+	if id == "" {
+		log.ErrorR(req, fmt.Errorf("payment id not supplied"))
 		w.WriteHeader(http.StatusBadRequest)
 		return
 	}
 
-	requestDecoder := json.NewDecoder(req.Body)
-	var incomingExternalPaymentJourneyRequest models.IncomingExternalPaymentJourneyRequest
-	err := requestDecoder.Decode(&incomingExternalPaymentJourneyRequest)
+	paymentSession, httpStatus, err := (*PaymentService).getPaymentSession(service, id)
```
Comment (jterrych) [inline service/external_paysession.go]: Is there any reason why the method has to be called this way? isn't it more readable to call `service.getPaymentSession(id)` instead?

Comment (jphillips2121) [reply service/external_paysession.go]: There wasn't any reason. Changed now.

Comment (jterrych) [reply service/external_paysession.go]: 👍

File: service/gov_pay.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+	if err != nil {
+		return "", fmt.Errorf("error reading response from GovPay: [%s]", err)
+	}
+
+	govPayResponse := &models.IncomingGovPayResponse{}
+	err = json.Unmarshal(body, govPayResponse)
+	if err != nil {
+		return "", fmt.Errorf("error reading response from GovPay: [%s]", err)
+	}
+	if resp.StatusCode != 201 {
```
Comment (divya051988) [inline service/gov_pay.go]: Could you please use http.StatusCreated here instead of 201 ?

Comment (jphillips2121) [reply service/gov_pay.go]: Done

File: service/external_paysession_test.go [hunk 4]
```diff
@@ -1,52 +1,183 @@
...
+		w := httptest.NewRecorder()
+
+		httpmock.Activate()
+		defer httpmock.DeactivateAndReset()
+		costArray := []models.CostResource{defaultCost}
+		jsonResponse, _ := httpmock.NewJsonResponder(200, costArray)
+		httpmock.RegisterResponder("GET", "http://dummy-resource", jsonResponse)
+
+		IncomingGovPayResponse := models.IncomingGovPayResponse{}
+		govPayJSONResponse, _ := httpmock.NewJsonResponder(201, IncomingGovPayResponse)
```
Comment (divya051988) [inline service/external_paysession_test.go]: http.statusCreated could be used here as well.

Comment (jphillips2121) [reply service/external_paysession_test.go]: Changed both of them

Comment (divya051988) [reply service/external_paysession_test.go]: https://github.com/companieshouse/payments.api.ch.gov.uk/pull/44/files#diff-1ef3aad51fb2ffb3106ab2df065ea7b2R80

Comment (jphillips2121) [reply service/external_paysession_test.go]: Fixed

**Pull request merged**